### DISCUSSION TITLE
Add a node to unproject 3D position to 2D

### DISCUSSION THREAD

**Discussion by timothyqiu** [godotengine/godot-proposals#5683] [3D]
It's trivial to add something like this in `_process()`:

```gdscript
var cam := get_viewport().get_camera()
var pos := get_parent().global_translation
rect_position = cam.unproject_position(pos)
```

I think it could be helpful to add a `Unprojector2D` that extends `Node2D`. When its parent is a `Node3D`, it updates its position the match the parent's unprojected position.

**Comment by Calinou** [comment]
I remember seeing a proposal about adding a new RemoteTransform node to cross the boundary between 2D and 3D. That said, I don't think it's a common use case enough to warrant being a core feature – this can be done with an add-on.

**Reply by timothyqiu** [reply]
The use case is like [3D Waypoints Demo](https://github.com/godotengine/godot-demo-projects/tree/master/3d/waypoints). It's not for 2D GUI in 3D, but matching 2D annotation elements for 3D objects like health bars, name plates, and action key prompts.

Viewports have a performance cost. So instead of using a billboard & fixed size Sprite3D, it's more acceptable to use plain 2D. Label3D solved this problem for texts, but the problem remains when other Controls are needed.

**Comment by Zireael07** [comment]
Counter argument: Not everything has to be a node, they have a perf cost

**Reply by timothyqiu** [reply]
I remember in a multiple-script-per-node issue/proposal and in a virtual folder in tree proposal, it's recommended to add extra nodes to the tree instead since it has no performance cost.
